bpo-40059: tomllib (#2)
hukkin/cpython: Doc/library/tomllib.rst

support writing TOML.

.. seealso::

    The `Tomli-W package <https://pypi.org/project/tomli-w/>`__
    is a TOML writer that can be used in conjunction with this module,
    providing a write API familiar to users of the standard library
    :mod:`marshal` and :mod:`pickle` modules.

.. seealso::

    The `TOML Kit package <https://pypi.org/project/tomlkit/>`__
    is a style-preserving TOML library with both read and write capability.
    It is a recommended replacement for this module for editing already
    existing TOML files.

This module defines the following functions:

.. function:: load(fp, /, *, parse_float=float)

   Read a TOML file. The first argument should be a readable and binary file object.
   Return a :class:`dict`. Convert TOML types to Python using this
   :ref:`conversion table <toml-to-py-table>`.

   *parse_float* will be called with the string of every TOML
   float to be decoded.  By default, this is equivalent to ``float(num_str)``.
   This can be used to use another datatype or parser for TOML floats
   (e.g. :class:`decimal.Decimal`). The callable must not return a
   :class:`dict`, a :class:`list`, or anything that has the ``append``

>    attribute. These illegal types produce undefined behavior.

Reviewer: That's a weird limitation. Can it be lifted, or can tomllib raise a hard error instead of the undefined behavior?

Author: The `append` attr limitation is easy to lift. Basically just move from a `try-except` to an `isinstance` check, meaning a very small performance cost.

The `dict` and `list` limitation are not so simple at all, I wouldn't recommend trying to lift them. It should be fairly simple to make these hard errors though (at a small performance cost, that can be avoided when `parse_float` is set to default value).

Author: ```suggestion
   (e.g. :class:`decimal.Decimal`). The callable must not return a
   :class:`dict`, a :class:`list`, else a `ValueError` is raised.
```
Would this be a more acceptable limitation?

Reviewer: Definitely. A `TypeError` might be more appropriate.

Author: Hmm, are you sure about that. If the callable returns something then it's still... a callable. And that's the type annotated requirement. If it returns `dict` or `list` then I'd see the callable as appropriate type but inappropriate value.

Author: Wonder if there's a precedent of something similar in the stdlib?

Reviewer 2: TypeError seems fine.  You want a dict or list and you don’t get that, so a type is not good 🙂

Author: Sure that's true but didn't the real error happen already before that, when the user input incorrect **value** to `parse_float`?

Haha, this isn't that simple :smile:

Reviewer: If the error was reported there, sure. But AFAICS, tomllib can only check the result of the callable -- and it checks if the type is correct.
Not sure if there's a precedent. But it's a minor nitpick, if you disagree then just close the conversation :)

Author: > If the error was reported there, sure. But AFAICS, tomllib can only check the result of the callable -- and it checks if the type is correct.

Yeh you're right, but from the user perspective, there's no difference where the error is reported, whether it's right after calling `tomllib.loads` or only when the first `float` is found. It's all within the same function call.

According to Python docs:
> [TypeError is] Raised when an operation or function is applied to an object of inappropriate type.

But we don't apply any operations or functions to the invalid type created by `parse_float`. That is not our error. The error is that invalid `parse_float` value created that type in the first place.

Based on this reasoning I'll start with `ValueError`, but yeh, it's not like I'm definitely 100% sure about this. :smile:  I'm still more than willing to change my stance given good reasoning.

Reviewer 2: What I wanted to say is that there isn’t an easy, strong, perfect rule for this, and that it does not matter much 🙂

:mod:`tomllib` --- Parse TOML files
===================================

.. module:: tomllib
   :synopsis: Parse TOML files.

.. moduleauthor:: TODO-firstname TODO-lastname

> .. sectionauthor:: TODO-firstname TODO-lastname <todo@todo.org>

Author: :smile: Honestly, I was unsure whether these can be non CPython core dev names? I.e. should these be me or Petr Viktorin?

> .. sectionauthor:: TODO-firstname TODO-lastname <todo@todo.org>

Reviewer: They can be anything. But they can also be removed entirely.

> .. sectionauthor:: TODO-firstname TODO-lastname <todo@todo.org>

Reviewer 2: They should be the author of tomllib and the author of the docs.

.. exception:: TOMLDecodeError

   Subclass of :exc:`ValueError`.

Examples
--------

Parsing a TOML file::

    import tomllib

    with open("pyproject.toml", "rb") as f:
        data = tomllib.load(f)

Parsing a TOML string::

    import tomllib

>     toml_str = """

Reviewer 2: load wants a binary stream but loads a text string?

Author: That's correct: https://www.python.org/dev/peps/pep-0680/#id18
